File: golang/internal/trading/executor.go
package trading

import (
    "errors"

    "github.com/example/crypto_arbitrage_monitor/golang/internal/arbitrage"
    "github.com/example/crypto_arbitrage_monitor/golang/internal/monitors"
    "github.com/example/crypto_arbitrage_monitor/golang/internal/risk"
    "github.com/example/crypto_arbitrage_monitor/golang/internal/utils"
)

type DexTrader struct {
    Name string
    Fee  float64
}

type CexTrader struct {
    Name string
    Fee  float64
}

type Executor struct {
    Calculator   *arbitrage.Calculator
    DexTrader    DexTrader
    CexTrader    CexTrader
    RiskManager  *risk.Manager
    DexExchanges map[string]struct{}
    CexExchanges map[string]struct{}
}

type TradeResult struct {
    Opportunity arbitrage.Opportunity
    Amount      float64
    Profit      float64
    Status      string
}

func (t DexTrader) Buy(amount, price float64) float64 {
    return amount * price * (1 + t.Fee)
}

func (t DexTrader) Sell(amount, price float64) float64 {
    return amount * price * (1 - t.Fee)
}

func (t CexTrader) Buy(amount, price float64) float64 {
    return amount * price * (1 + t.Fee)
}

func (t CexTrader) Sell(amount, price float64) float64 {
    return amount * price * (1 - t.Fee)
}

func (e *Executor) Execute(opportunity arbitrage.Opportunity, priceMap map[string]monitors.PricePoint) (*TradeResult, error) {
    buyPoint, ok := priceMap[opportunity.BuyExchange]
    if !ok {
        return nil, errors.New("missing buy price")
    }
    sellPoint, ok := priceMap[opportunity.SellExchange]
    if !ok {
        return nil, errors.New("missing sell price")
    }

    available := e.RiskManager.AvailableCapital()
    amount := available * opportunity.Confidence
    if amount <= 0 {
        return nil, errors.New("no available capital")
    }
    if !e.RiskManager.Allocate(opportunity.ID, amount) {
        return nil, errors.New("risk manager rejected trade")
    }

    breakdown := e.Calculator.Evaluate(buyPoint.Price, sellPoint.Price, amount)
    if breakdown.NetProfit <= 0 {
        e.RiskManager.Release(opportunity.ID, amount, 0)
        return nil, errors.New("negative profit")
    }

    cost, err := e.executeBuy(opportunity.BuyExchange, amount, buyPoint.Price)
    if err != nil {
        e.RiskManager.Release(opportunity.ID, amount, 0)
        return nil, err
    }
    revenue, err := e.executeSell(opportunity.SellExchange, amount, sellPoint.Price)
    if err != nil {
        e.RiskManager.Release(opportunity.ID, amount, 0)
        return nil, err
    }
    profit := revenue - cost
    e.RiskManager.Record(opportunity.ID, amount, profit)
    utils.Logger().Printf("Executed %s profit %.4f", opportunity.Pair, profit)
    return &TradeResult{Opportunity: opportunity, Amount: amount, Profit: profit, Status: "executed"}, nil
}

func (e *Executor) executeBuy(exchange string, amount, price float64) (float64, error) {
    if _, ok := e.DexExchanges[exchange]; ok {
        return e.DexTrader.Buy(amount, price), nil
    }
    if _, ok := e.CexExchanges[exchange]; ok {
        return e.CexTrader.Buy(amount, price), nil
    }
    return 0, errors.New("unknown exchange")
}

func (e *Executor) executeSell(exchange string, amount, price float64) (float64, error) {
    if _, ok := e.DexExchanges[exchange]; ok {
        return e.DexTrader.Sell(amount, price), nil
    }
    if _, ok := e.CexExchanges[exchange]; ok {
        return e.CexTrader.Sell(amount, price), nil
    }
    return 0, errors.New("unknown exchange")
}

File: src/trading/executor.py
"""Trade execution orchestration."""
from __future__ import annotations

from dataclasses import dataclass
from typing import Dict

from ..arbitrage.calculator import OpportunityCalculator
from ..arbitrage.strategies import Strategy
from ..risk.manager import RiskManager
from ..utils.database import InMemoryDatabase, StoredTrade
from ..utils.logger import get_logger
from ..utils.models import ArbitrageOpportunity, PricePoint, TradeResult
from ..utils.notifications import Notifier
from .cex_trader import CexTrader
from .dex_trader import DexTrader

logger = get_logger(__name__)


@dataclass
class TradeExecutor:
    calculator: OpportunityCalculator
    dex_trader: DexTrader
    cex_trader: CexTrader
    risk_manager: RiskManager
    notifier: Notifier
    database: InMemoryDatabase
    strategy: Strategy
    dex_exchanges: set[str]
    cex_exchanges: set[str]

    def execute(self, opportunity: ArbitrageOpportunity, price_map: Dict[str, PricePoint]) -> TradeResult | None:
        buy_point = price_map.get(opportunity.buy_exchange)
        sell_point = price_map.get(opportunity.sell_exchange)
        if not buy_point or not sell_point:
            logger.debug("Missing prices for opportunity %s", opportunity.identifier)
            return None

        available_capital = self.risk_manager.available_capital
        trade_amount = self.strategy.plan_trade(opportunity, available_capital)
        if trade_amount <= 0:
            logger.debug("Strategy returned zero trade size for %s", opportunity.identifier)
            return None
        if not self.risk_manager.allocate(opportunity.identifier, trade_amount):
            return None

        breakdown = self.calculator.evaluate(buy_point.price, sell_point.price, trade_amount)
        if breakdown.net_profit <= 0:
            logger.info("Opportunity %s rejected due to negative net profit", opportunity.identifier)
            self.risk_manager.portfolio.release(opportunity.identifier, trade_amount, 0.0)
            return None

        buy_cost = self._execute_buy(opportunity.buy_exchange, opportunity.pair, trade_amount, buy_point.price)
        sell_revenue = self._execute_sell(opportunity.sell_exchange, opportunity.pair, trade_amount, sell_point.price)
        realized_profit = sell_revenue - buy_cost

        trade = TradeResult(
            opportunity_id=opportunity.identifier,
            executed_amount=trade_amount,
            realized_profit=round(realized_profit, 4),
            status="executed" if realized_profit > 0 else "breakeven",
        )
        self.risk_manager.register_trade(opportunity.identifier, trade_amount, realized_profit)
        self.database.add_trade(
            StoredTrade(
                opportunity_id=trade.opportunity_id,
                amount=trade.executed_amount,
                realized_profit=trade.realized_profit,
                status=trade.status,
            )
        )
        self.notifier.notify(
            f"Executed arbitrage {opportunity.pair} profit {trade.realized_profit:.4f}"
        )
        return trade

    def _execute_buy(self, exchange: str, pair: str, amount: float, price: float) -> float:
        if exchange in self.dex_exchanges:
            return self.dex_trader.buy(pair, amount, price)
        if exchange in self.cex_exchanges:
            return self.cex_trader.buy(pair, amount, price)
        raise ValueError(f"Unknown exchange for buy: {exchange}")

    def _execute_sell(self, exchange: str, pair: str, amount: float, price: float) -> float:
        if exchange in self.dex_exchanges:
            return self.dex_trader.sell(pair, amount, price)
        if exchange in self.cex_exchanges:
            return self.cex_trader.sell(pair, amount, price)
        raise ValueError(f"Unknown exchange for sell: {exchange}")
